File: Binary_Search_Trees.js
// Node class representing each node in the BST
class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

// Tree class representing the binary search tree
class Tree {
    constructor(array) {
        this.root = this.buildTree(array);
    }

    // Function to build a balanced binary search tree from an array
    buildTree(array) {
        if (array.length === 0) return null;

        const sortedArray = Array.from(new Set(array)).sort((a, b) => a - b);

        const buildSubTree = (arr, start, end) => {
            if (start > end) return null;

            const mid = Math.floor((start + end) / 2);
            const node = new Node(arr[mid]);

            node.left = buildSubTree(arr, start, mid - 1);
            node.right = buildSubTree(arr, mid + 1, end);

            return node;
        };

        return buildSubTree(sortedArray, 0, sortedArray.length - 1);
    }

    // Function to insert a value into the BST
    insert(value) {
        const insertNode = (node, value) => {
            if (node === null) return new Node(value);

            if (value < node.data) {
                node.left = insertNode(node.left, value);
            } else if (value > node.data) {
                node.right = insertNode(node.right, value);
            }

            return node;
        };

        this.root = insertNode(this.root, value);
    }

    // Function to delete a value from the BST
    delete(value) {
        const deleteNode = (node, value) => {
            if (node === null) return null;

            if (value < node.data) {
                node.left = deleteNode(node.left, value);
            } else if (value > node.data) {
                node.right = deleteNode(node.right, value);
            } else {
                if (node.left === null && node.right === null) {
                    return null;
                } else if (node.left === null) {
                    return node.right;
                } else if (node.right === null) {
                    return node.left;
                }

                const minRight = findMin(node.right);
                node.data = minRight.data;
                node.right = deleteNode(node.right, minRight.data);
            }

            return node;
        };

        const findMin = (node) => {
            while (node.left !== null) {
                node = node.left;
            }
            return node;
        };

        this.root = deleteNode(this.root, value);
    }

    // Function to find a node with a given value
    find(value) {
        const findNode = (node, value) => {
            if (node === null) return null;

            if (value === node.data) {
                return node;
            } else if (value < node.data) {
                return findNode(node.left, value);
            } else {
                return findNode(node.right, value);
            }
        };

        return findNode(this.root, value);
    }

    // Function to traverse the tree in breadth-first level order
    levelOrder(callback) {
        if (!this.root) return [];

        const queue = [this.root];
        const result = [];

        while (queue.length > 0) {
            const node = queue.shift();
            result.push(node.data);

            if (node.left !== null) {
                queue.push(node.left);
            }

            if (node.right !== null) {
                queue.push(node.right);
            }

            if (callback) {
                callback(node);
            }
        }

        return result;
    }

    // Function to perform in-order traversal of the tree
    inOrder(callback) {
        const traverse = (node) => {
            if (node !== null) {
                traverse(node.left);
                if (callback) callback(node);
                traverse(node.right);
            }
        };

        traverse(this.root);
    }

    // Function to perform pre-order traversal of the tree
    preOrder(callback) {
        const traverse = (node) => {
            if (node !== null) {
                if (callback) callback(node);
                traverse(node.left);
                traverse(node.right);
            }
        };

        traverse(this.root);
    }

    // Function to perform post-order traversal of the tree
    postOrder(callback) {
        const traverse = (node) => {
            if (node !== null) {
                traverse(node.left);
                traverse(node.right);
                if (callback) callback(node);
            }
        };

        traverse(this.root);
    }

    // Function to calculate the height of a given node
    height(node) {
        if (node === null) return -1;

        const leftHeight = this.height(node.left);
        const rightHeight = this.height(node.right);

        return Math.max(leftHeight, rightHeight) + 1;
    }

    // Function to calculate the depth of a given node
    depth(node) {
        if (node === null) return 0;

        let depth = 0;
        let currentNode = node;

        while (currentNode !== null) {
            depth++;
            currentNode = currentNode.parent;
        }

        return depth;
    }

    // Function to check if the tree is balanced
    isBalanced() {
        const checkBalanced = (node) => {
            if (node === null) return true;

            const leftHeight = this.height(node.left);
            const rightHeight = this.height(node.right);

            if (Math.abs(leftHeight - rightHeight) <= 1 &&
                checkBalanced(node.left) &&
                checkBalanced(node.right)) {
                return true;
            }

            return false;
        };

        return checkBalanced(this.root);
    }

    // Function to rebalance the tree
    rebalance() {
        const nodes = [];
        this.inOrder((node) => nodes.push(node.data));

        this.root = this.buildTree(nodes);
    }
}

// Function to print a tree structure
const prettyPrint = (node, prefix = "", isLeft = true) => {
    if (node === null) {
        return;
    }
    if (node.right !== null) {
        prettyPrint(node.right, `${prefix}${isLeft ? "│   " : "    "}`, false);
    }
    console.log(`${prefix}${isLeft ? "└── " : "┌── "}${node.data}`);
    if (node.left !== null) {
        prettyPrint(node.left, `${prefix}${isLeft ? "    " : "│   "}`, true);
    }
};

// Driver script
const getRandomNumbers = (count) => {
    const numbers = new Set();
    while (numbers.size < count) {
        numbers.add(Math.floor(Math.random() * 100));
    }
    return Array.from(numbers);
};

const bst = new Tree(getRandomNumbers(15));
console.log("Is balanced:", bst.isBalanced());
console.log("Level order:");
console.log(bst.levelOrder());
console.log("Pre order:");
bst.preOrder((node) => console.log(node.data));
console.log("Post order:");
bst.postOrder((node)

File: binary_search_trees.rb
# Node class for creating nodes in the tree
class Node
  include Comparable

  attr_accessor :data, :left, :right

  def initialize(data)
    @data = data
    @left = nil
    @right = nil
  end

  def <=>(other)
    @data <=> other.data
  end
end

# Tree class for creating and managing the binary search tree
class Tree
  attr_accessor :root

  def initialize(array)
    @root = build_tree(array.uniq.sort)
  end

  # Build a balanced binary search tree from a sorted array
  def build_tree(array)
    return nil if array.empty?

    mid = (array.length - 1) / 2
    root = Node.new(array[mid])

    root.left = build_tree(array[0...mid])
    root.right = build_tree(array[mid + 1..-1])

    root
  end

  # Insert a value into the tree
  def insert(value, node = @root)
    return Node.new(value) unless node

    if value < node.data
      node.left = insert(value, node.left)
    elsif value > node.data
      node.right = insert(value, node.right)
    end

    node
  end

  # Delete a value from the tree
  def delete(value, node = @root)
    return nil unless node

    if value < node.data
      node.left = delete(value, node.left)
    elsif value > node.data
      node.right = delete(value, node.right)
    else
      return node.right unless node.left
      return node.left unless node.right

      temp = min_value_node(node.right)
      node.data = temp.data
      node.right = delete(temp.data, node.right)
    end

    node
  end

  # Find a node with a given value
  def find(value, node = @root)
    return nil unless node

    if value < node.data
      find(value, node.left)
    elsif value > node.data
      find(value, node.right)
    else
      node
    end
  end

  # Perform a level order traversal of the tree
  def level_order
    return [] unless @root

    queue = [@root]
    result = []

    until queue.empty?
      current = queue.shift
      result << current.data

      queue << current.left if current.left
      queue << current.right if current.right
    end

    block_given? ? result.each { |data| yield data } : result
  end

  # Perform an inorder traversal of the tree
  def inorder(node = @root, &block)
    return [] unless node

    inorder(node.left, &block) + [node.data] + inorder(node.right, &block)
  end

  # Perform a preorder traversal of the tree
  def preorder(node = @root, &block)
    return [] unless node

    [node.data] + preorder(node.left, &block) + preorder(node.right, &block)
  end

  # Perform a postorder traversal of the tree
  def postorder(node = @root, &block)
    return [] unless node

    postorder(node.left, &block) + postorder(node.right, &block) + [node.data]
  end

  # Get the height of a node in the tree
  def height(node = @root)
    return -1 unless node

    [height(node.left), height(node.right)].max + 1
  end

  # Get the depth of a node in the tree
  def depth(value, node = @root)
    return 0 unless node

    if value < node.data
      1 + depth(value, node.left)
    elsif value > node.data
      1 + depth(value, node.right)
    else
      0
    end
  end

  # Check if the tree is balanced
  def balanced?(node = @root)
    return true unless node

    left_height = height(node.left)
    right_height = height(node.right)
    (left_height - right_height).abs <= 1 && balanced?(node.left) && balanced?(node.right)
  end

  # Rebalance the tree
  def rebalance
    @root = build_tree(inorder)
  end

  private

  # Get the minimum value node in the tree
  def min_value_node(node)
    current = node
    current = current.left while current.left
    current
  end
end

# Driver script
def driver_script
  # Create a binary search tree from an array of random numbers
  array = Array.new(15) { rand(1..100) }
  tree = Tree.new(array)

  puts 'Original Tree:'
  puts '---------------------'
  puts 'Balanced? :', tree.balanced?
  puts 'Level order:', tree.level_order
  puts 'Preorder:', tree.preorder
  puts 'Inorder:', tree.inorder
  puts 'Postorder:', tree.postorder
  puts '---------------------'

  # Unbalance the tree by adding several numbers > 100
  array = Array.new(10) { rand(101..200) }
  array.each { |num| tree.insert(num) }

  puts 'Tree after unbalancing:'
  puts '---------------------'
  puts 'Balanced? :', tree.balanced?
  puts 'Level order:', tree.level_order
  puts 'Preorder:', tree.preorder
  puts 'Inorder:', tree.inorder
  puts 'Postorder:', tree.postorder
  puts '---------------------'

  # Balance the tree
  tree.rebalance

  puts 'Tree after rebalancing:'
  puts '---------------------'
  puts 'Balanced? :', tree.balanced?
  puts 'Level order:', tree.level_order
  puts 'Preorder:', tree.preorder
  puts 'Inorder:', tree.inorder
  puts 'Postorder:', tree.postorder
  puts '---------------------'
end

# Run the driver script
driver_script if __FILE__ == $PROGRAM_NAME
